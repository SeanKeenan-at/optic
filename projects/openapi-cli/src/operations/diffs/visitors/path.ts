import { OpenAPIV3 } from '../../../specs';
import { OperationDiffResult, OperationDiffResultKind } from '../result';
import { PathComponents, PathComponent } from '../..';
import { Option } from 'ts-results';

export function* visitPath(
  path: string,
  specOption: Option<OpenAPIV3.PathItemObject>,
  context: { pathPattern: Option<string> }
): IterableIterator<OperationDiffResult> {
  let parameterComponents = PathComponents.fromPath(path).filter(
    PathComponent.isTemplate
  );

  if (specOption.none) {
    yield {
      kind: OperationDiffResultKind.UnmatchedPath,
      subject: path,
    };

    for (let parameterComponent of parameterComponents) {
      yield {
        kind: OperationDiffResultKind.UnmatchedPathParameter,
        subject: parameterComponent.name,
      };
    }
  }

  // TODO: match path's parameters against those documented
  // const spec = specOption.unwrap()
  // spec.parameters
}
