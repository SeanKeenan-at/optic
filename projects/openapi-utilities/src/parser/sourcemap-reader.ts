import {
  DerefToSource,
  JsonPath,
  JsonSchemaSourcemap,
  resolveJsonPointerInYamlAst,
  ToSource,
} from "./openapi-sourcemap-parser";
import fs from "fs-extra";
import { Kind, YamlMap, YAMLNode, YAMLSequence } from "yaml-ast-parser";
import { jsonPointerHelper } from "../index";
import equals from "fast-deep-equal";

export function sourcemapReader(sourcemap: JsonSchemaSourcemap) {
  const rootFileNumber = sourcemap.files.find(
    (i) => i.path === sourcemap.rootFilePath
  )!.index;

  const findFile = (
    jsonPathFromRoot: JsonPath,
    currentFile: number = rootFileNumber
  ): ILookupPathResult => {
    const decoded = jsonPointerHelper.decode(jsonPathFromRoot);

    // console.log("running it", jsonPathFromRoot, currentFile);

    let consideredRefs = sourcemap.refMappings;

    let pathStartLookup: ToSource | undefined;

    decoded.forEach((component, index) => {
      const filtered = consideredRefs.filter((trailMappings) => {
        const [flatTrail] = trailMappings;
        const atLength = flatTrail.slice(0, index + 1);
        return equals(atLength, decoded.slice(0, index + 1));
      });
      if (consideredRefs.length > 0 && filtered.length === 0) {
        pathStartLookup = consideredRefs[0];
      }
      consideredRefs = filtered;
    });

    // console.log(pathStartLookup);

    if (pathStartLookup) {
      const [pathInRoot, fileN, jsonPathInFile] = pathStartLookup;
      const pointerFull = jsonPointerHelper.compile([
        ...jsonPathInFile,
        ...decoded.slice(pathInRoot.length),
      ]);

      const fileLookup = sourcemap.files.find((i) => i.index === fileN)!;

      if (fileLookup.index === currentFile) {
        return findFile(pointerFull, currentFile);
      }

      const node = resolveJsonPointerInYamlAst(fileLookup.ast, pointerFull);

      if (node) {
        return {
          filePath: fileLookup.path,
          astNode: node,
          startsAt: jsonPointerHelper.compile(jsonPathInFile),
        };
      }
    } else {
      // assume root file
      const fileLookup = sourcemap.files.find((i) => i.index === currentFile);
      if (fileLookup) {
        const node = resolveJsonPointerInYamlAst(
          fileLookup.ast,
          jsonPathFromRoot
        );
        if (node) {
          return {
            filePath: fileLookup.path,
            astNode: node,
            startsAt: jsonPathFromRoot,
          };
        }
      }
    }
  };

  const findFileAndLines = async (jsonPathFromRoot: JsonPath) => {
    const lookupResult = findFile(jsonPathFromRoot);
    if (lookupResult) {
      const astNode = lookupResult.astNode;
      const contents = (await fs.readFile(lookupResult.filePath)).toString();

      const [startPosition, endPosition] = astNodesToStartEndPosition(astNode);

      const { startLine, endLine, preview } = positionToLine(
        contents,
        startPosition,
        endPosition
      );
      const result: ILookupLinePreviewResult = {
        filePath: lookupResult.filePath,
        startLine,
        endLine,
        preview,
        startPosition: startPosition,
        endPosition: endPosition,
      };
      return result;
    }
  };

  return {
    findFile,
    findFileAndLines,
  };
}

type ILookupPathResult =
  | undefined
  | { filePath: string; startsAt: JsonPath; astNode: YAMLNode };
export type ILookupLinePreviewResult =
  | undefined
  | {
      endLine: number;
      endPosition: number;
      filePath: string;
      startLine: number;
      preview: string;
      startPosition: number;
    };

//////////////////////////////////////////////////////////

function positionToLine(
  contents: string,
  start: number,
  end: number
): { startLine: number; endLine: number; preview: string } {
  const startLine =
    (contents.substring(0, start).match(/\n/g) || "").length + 1;
  const endLine =
    (contents.substring(end).match(/\n/g) || "").length + startLine - 1;

  const lines = contents.split(/\r\n|\r|\n/);

  const preview = lines.slice(startLine - 1, endLine).join("\n");

  return {
    startLine,
    endLine,
    preview,
  };
}

function astNodesToStartEndPosition(astNode: YAMLNode): [number, number] {
  switch (astNode.kind) {
    case Kind.MAP: {
      const map = astNode as YamlMap;
      const end =
        map.value.mappings[map.value.mappings.length - 1]?.endPosition ||
        astNode.endPosition;
      return [map.startPosition, end];
    }
    case Kind.SEQ: {
      const seq = astNode as YAMLSequence;
      const end =
        seq.items[seq.items.length - 1]?.endPosition || astNode.endPosition;
      return [seq.startPosition, end];
    }
    default:
      return [astNode.startPosition, astNode.endPosition];
  }
}
