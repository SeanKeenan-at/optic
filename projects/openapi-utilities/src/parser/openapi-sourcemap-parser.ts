import $RefParser from "@apidevtools/json-schema-ref-parser";
// @ts-ignore
import * as $RefParserOptions from "@apidevtools/json-schema-ref-parser/lib/options";
import * as YAML from "yaml-ast-parser";
import * as fs from "fs-extra";
import { YAMLMapping, YAMLNode, YAMLSequence } from "yaml-ast-parser";
// @ts-ignore
import { dereference } from "./insourced-dereference";
import * as pointer from "json-ptr";
import path from "path";
import fetch from "node-fetch";
import { OpenAPIV3 } from "openapi-types";
import sortBy from "lodash.sortby";

export type ParseOpenAPIResult = {
  jsonLike: OpenAPIV3.Document;
  sourcemap: JsonSchemaSourcemapOutput;
};

export async function parseOpenAPIWithSourcemap(
  path: string
): Promise<ParseOpenAPIResult> {
  const resolver = new $RefParser();

  const sourcemap = new JsonSchemaSourcemap();
  const resolverResults: $RefParser.$Refs = await resolver.resolve(path, {
    resolve: {
      http: {
        headers: {
          accept: "*/*",
        },
      },
    },
  });

  // parse all asts
  await Promise.all(
    resolverResults
      .paths()
      .map((filePath, index) => sourcemap.addFileIfMissing(filePath, index))
  );

  dereference(
    resolver,
    { ...$RefParserOptions.defaults, path: path },
    sourcemap
  );

  return { jsonLike: resolver.schema as any, sourcemap: sourcemap.serialize() };
}

export async function parseOpenAPIFromRepoWithSourcemap(
  name: string,
  repoPath: string,
  branch: string
): Promise<ParseOpenAPIResult> {
  const newGitBranchResolver = require("./git-branch-file-resolver.js");

  const inGitResolver = newGitBranchResolver(repoPath, branch);
  const resolver = new $RefParser();
  const fileName = path.join(repoPath, name);

  const sourcemap = new JsonSchemaSourcemap();
  const resolverResults: $RefParser.$Refs = await resolver.resolve(fileName, {
    resolve: { file: inGitResolver },
  });

  // parse all asts
  await Promise.all(
    resolverResults.paths().map(async (filePath, index) => {
      return await sourcemap.addFileIfMissingFromContents(
        filePath,
        await inGitResolver.read({ url: filePath }),
        index
      );
    })
  );

  dereference(
    resolver,
    {
      ...$RefParserOptions.defaults,
      path: fileName,
      resolve: { file: inGitResolver },
    },
    sourcemap
  );

  return { jsonLike: resolver.schema as any, sourcemap: sourcemap.serialize() };
}

type JsonPath = string;
type FileReference = number;

type DerefToSource = [JsonPath, LocationRecord];

export interface JsonSchemaSourcemapOutput {
  files: Array<{
    path: string;
    index: number;
  }>;
  map: DerefToSource[];
}

export class JsonSchemaSourcemap {
  private _files: Array<{
    path: string;
    index: number;
    ast: YAMLNode;
  }> = [];

  private _mappings: Array<DerefToSource> = [];

  async addFileIfMissing(filePath: string, fileIndex: number) {
    if (filePath.startsWith("http")) {
      const response = await fetch(filePath);
      const asText = await response.text();

      const yamlAst: YAMLNode = YAML.safeLoad(asText);

      this._files.push({
        path: filePath,
        index: fileIndex,
        ast: yamlAst,
      });
    } else {
      if (!this._files.find((i) => i.path === filePath)) {
        // add the ast to the cache
        const yamlAst: YAMLNode = YAML.safeLoad(
          (await fs.readFile(filePath)).toString()
        );

        this._files.push({
          path: filePath,
          index: fileIndex,
          ast: yamlAst,
        });
      }
    }
  }

  async addFileIfMissingFromContents(
    filePath: string,
    contents: string,
    fileIndex: number
  ) {
    if (!this._files.find((i) => i.path === filePath)) {
      // add the ast to the cache
      const yamlAst: YAMLNode = YAML.safeLoad(contents);

      this._files.push({
        path: filePath,
        index: fileIndex,
        ast: yamlAst,
      });
    }
  }

  log(path: string, pathFromRoot: string) {
    // this seems to assume that paths will be in order, why not check for equality?
    const thisFile = this._files.find((i) => path.startsWith(i.path));
    if (thisFile) {
      const jsonPointer = path.split(thisFile.path)[1].substring(1) || "/";
      const sourceMapping = resolveJsonPointerInYamlAst(
        thisFile.ast,
        jsonPointer,
        thisFile.index
      );
      if (sourceMapping) {
        this._mappings.push([pathFromRoot, sourceMapping]);
      }
    }
  }

  public serialize(): JsonSchemaSourcemapOutput {
    return {
      files: sortBy(
        this._files.map((i) => ({ path: i.path, index: i.index })),
        "index"
      ),
      map: this._mappings,
    };
  }
}

export function resolveJsonPointerInYamlAst(
  node: YAMLNode,
  jsonPointer: string,
  file: number
): LocationRecord | undefined {
  const decoded = pointer.decodePointer(jsonPointer);

  const isEmpty =
    decoded.length === 0 || (decoded.length === 1 && decoded[0] === "");

  if (isEmpty)
    return { node: [node.startPosition, node.endPosition], file: file };

  const found: YAMLNode | undefined = decoded.reduce((current, path) => {
    if (!current) return undefined;
    const node: YAMLNode = current.key ? current.value : current;
    const isNumericalKey =
      !isNaN(Number(path)) && (node as any).hasOwnProperty("items");

    if (isNumericalKey) {
      return (node as YAMLSequence).items[Number(path)];
    } else {
      const field = node.mappings.find(
        (i: YAMLMapping) => i.key.value === path
      );
      return field;
    }
  }, node as YAMLNode | undefined);

  if (found) {
    if (found.key) {
      // is a field
      return {
        key: [found.key.startPosition, found.key.endPosition],
        value: [found.value.startPosition, found.value.endPosition],
        node: [found.startPosition, found.endPosition],
        file: file,
      };
    } else {
      return { node: [found.startPosition, found.endPosition], file: file };
    }
  }
}

type AstLocation = [number, number];
interface LocationRecord {
  key?: AstLocation;
  value?: AstLocation;
  node: AstLocation;
  file: number;
}
