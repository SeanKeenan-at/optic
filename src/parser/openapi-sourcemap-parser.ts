import $RefParser from "@apidevtools/json-schema-ref-parser";
// @ts-ignore
import * as $RefParserOptions from "@apidevtools/json-schema-ref-parser/lib/options";
import * as YAML from "yaml-ast-parser";
import * as fs from "fs-extra";
import { YAMLMapping, YAMLNode, YAMLSequence } from "yaml-ast-parser";
// @ts-ignore
import { dereference } from "./insourced-dereference";
import * as pointer from "json-ptr";
const newGitBranchResolver = require("./git-branch-file-resolver")
import path from 'path'


export async function parseOpenAPIWithSourcemap(path: string) {
  const resolver = new $RefParser();

  const sourcemap = new JsonSchemaSourcemap();
  const resolverResults: $RefParser.$Refs = await resolver.resolve(path);

  // parse all asts
  await Promise.all(
    resolverResults
      .paths()
      .map((filePath) => sourcemap.addFileIfMissing(filePath))
  );

  dereference(
    resolver,
    { ...$RefParserOptions.defaults, path: path },
    sourcemap
  );

  return { jsonLike: resolver.schema as any, sourcemap: sourcemap.serialize() };
}

export async function parseOpenAPIFromRepoWithSourcemap(name: string, repoPath: string, branch: string) {
  const inGitResolver = newGitBranchResolver(repoPath, branch)
  const resolver = new $RefParser();
  const fileName = path.join(repoPath, name)

  const sourcemap = new JsonSchemaSourcemap();
  const resolverResults: $RefParser.$Refs = await resolver.resolve(fileName, {resolve: {file: inGitResolver}});

  // parse all asts
  await Promise.all(
    resolverResults
      .paths()
      .map(async (filePath) => {
        return await sourcemap.addFileIfMissingFromContents(filePath, await inGitResolver.read({url: filePath}))
      })
  );

  dereference(
    resolver,
    { ...$RefParserOptions.defaults, path: fileName, resolve: {file: inGitResolver} },
    sourcemap
  );

  return { jsonLike: resolver.schema as any, sourcemap: sourcemap.serialize() };
}

type JsonPath = string;
type FileReference = number;

type DerefToSource = [JsonPath, LocationRecord];

export interface JsonSchemaSourcemapOutput {
  files: Array<{
    path: string;
    index: number;
  }>;
  map: DerefToSource[];
}

export class JsonSchemaSourcemap {
  private _files: Array<{
    path: string;
    index: number;
    ast: YAMLNode;
  }> = [];

  private _mappings: Array<DerefToSource> = [];

  async addFileIfMissing(filePath: string) {
    if (!this._files.find((i) => i.path === filePath)) {
      // add the ast to the cache
      const yamlAst: YAMLNode = YAML.safeLoad(
        (await fs.readFile(filePath)).toString()
      );

      this._files.push({
        path: filePath,
        index: this._files.length,
        ast: yamlAst,
      });
    }
  }

  async addFileIfMissingFromContents(filePath: string, contents: string) {
    if (!this._files.find((i) => i.path === filePath)) {
      // add the ast to the cache
      const yamlAst: YAMLNode = YAML.safeLoad(
        contents
      );

      this._files.push({
        path: filePath,
        index: this._files.length,
        ast: yamlAst,
      });
    }
  }

  log(path: string, pathFromRoot: string) {
    const thisFile = this._files.find((i) => path.startsWith(i.path));
    if (thisFile) {
      const jsonPointer = path.split(thisFile.path)[1].substring(1) || "/";
      // @todo remove this try catch, we want errors, but this is going to help us dev
      try {
        const sourceMapping = resolveJsonPointerInYamlAst(
          thisFile.ast,
          jsonPointer,
          thisFile.index
        );
        if (sourceMapping) {
          this._mappings.push([pathFromRoot, sourceMapping]);
        }
      } catch (e) {
        console.error("source mapping failure ", e)
      }
    }
  }

  public serialize(): JsonSchemaSourcemapOutput {
    return {
      files: this._files.map((i) => ({ path: i.path, index: i.index })),
      map: this._mappings,
    };
  }
}

export function resolveJsonPointerInYamlAst(
  node: YAMLNode,
  jsonPointer: string,
  file: number
): LocationRecord | undefined {
  const decoded = pointer.decodePointer(jsonPointer);

  const isEmpty =
    decoded.length === 0 || (decoded.length === 1 && decoded[0] === "");

  if (isEmpty) return { n: [node.startPosition, node.endPosition], f: file };

  const found: YAMLNode | undefined = decoded.reduce((current, path) => {
    const isFieldKey = isNaN(Number(path));
    if (!current) return undefined;

    const node: YAMLNode = current.key ? current.value : current;

    if (isFieldKey) {
      const field = node.mappings.find(
        (i: YAMLMapping) => i.key.value === path
      );
      return field;
    } else {
      // is number
      return (node as YAMLSequence).items[Number(path)];
    }
  }, node as YAMLNode | undefined);

  if (found) {
    if (found.key) {
      // is a field
      return {
        k: [found.key.startPosition, found.key.endPosition],
        v: [found.value.startPosition, found.value.endPosition],
        n: [found.startPosition, found.endPosition],
        f: file,
      };
    } else {
      return { n: [found.startPosition, found.endPosition], f: file };
    }
  }
}

type AstLocation = [number, number];
interface LocationRecord {
  k?: AstLocation;
  v?: AstLocation;
  n: AstLocation;
  f: number;
}
