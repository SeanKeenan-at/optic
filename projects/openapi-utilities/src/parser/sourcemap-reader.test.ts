import { parseOpenAPIWithSourcemap } from "./openapi-sourcemap-parser";
import path from "path";
import { sourcemapReader } from "./sourcemap-reader";

const fixture = async () => {
  return await parseOpenAPIWithSourcemap(
    path.resolve(
      path.join(
        __dirname,
        "../../inputs/openapi3-with-references/external-multiple.yaml"
      )
    )
  );
};

function stripCwd(file: string) {
  return file.replace(process.cwd(), "");
}

describe("reading sourcemaps", () => {
  it("can get file path of forign file for any pointer relative to flattened json", async () => {
    const results = await fixture();

    const node = sourcemapReader(results.sourcemap).findFile(
      "/properties/user/example/name"
    );

    expect(node!.filePath.endsWith("definitions.yaml")).toBeTruthy();
    expect(stripCwd(node!.filePath)).toMatchSnapshot();
  });

  it("can get file paths json pointer in main file relative to flattened json", async () => {
    const results = await fixture();

    const node = sourcemapReader(results.sourcemap).findFile("/properties");

    expect(node!.filePath.endsWith("external-multiple.yaml")).toBeTruthy();
    expect(stripCwd(node!.filePath)).toMatchSnapshot();
  });

  it("lines in affected file for sub-property", async () => {
    const results = await fixture();

    const node = await sourcemapReader(results.sourcemap).findFileAndLines(
      "/properties/user/example/name"
    );

    expect(node).toMatchSnapshot();
  });

  it("lines in affected for object", async () => {
    const results = await fixture();

    const node = await sourcemapReader(results.sourcemap).findFileAndLines(
      "/properties"
    );

    expect(node).toMatchSnapshot();
  });
});
