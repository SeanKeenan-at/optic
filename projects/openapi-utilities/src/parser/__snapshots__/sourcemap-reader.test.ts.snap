// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reading sourcemaps can get file path of forign file for any pointer relative to flattened json 1`] = `"/inputs/openapi3-with-references/definitions.yaml"`;

exports[`reading sourcemaps can get file paths json pointer in main file relative to flattened json 1`] = `"/inputs/openapi3-with-references/external-multiple.yaml"`;

exports[`reading sourcemaps lines in affected file for sub-property 1`] = `
Object {
  "endLine": 8,
  "endPosition": 111,
  "filePath": "/Users/aidancunniffe/Developer/openapi-optic/poc-governance-tools/projects/openapi-utilities/inputs/openapi3-with-references/definitions.yaml",
  "preview": "    name: 'Homer'",
  "startLine": 8,
  "startPosition": 98,
}
`;

exports[`reading sourcemaps lines in affected for object 1`] = `
Object {
  "endLine": 7,
  "endPosition": 117,
  "filePath": "/Users/aidancunniffe/Developer/openapi-optic/poc-governance-tools/projects/openapi-utilities/inputs/openapi3-with-references/external-multiple.yaml",
  "preview": "properties:
  token:
    type: string
  user:
    $ref: 'definitions.yaml#/User'",
  "startLine": 3,
  "startPosition": 37,
}
`;

exports[`reading sourcemaps resolve a large sampling of known-valid keys 1`] = `
Array [
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  Object {
    "endLine": 12,
    "endPosition": 292,
    "filePath": "/Users/aidancunniffe/Developer/openapi-optic/poc-governance-tools/projects/snyk-rules/end-end-tests/api-standards/components/responses/400.yaml",
    "preview": "",
    "startLine": 5,
    "startPosition": 207,
  },
  undefined,
  undefined,
  undefined,
  Object {
    "endLine": 34,
    "endPosition": 91,
    "filePath": "/Users/aidancunniffe/Developer/openapi-optic/poc-governance-tools/projects/snyk-rules/end-end-tests/api-standards/components/errors.yaml",
    "preview": "",
    "startLine": 4,
    "startPosition": 48,
  },
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  Object {
    "endLine": 1,
    "endPosition": 156,
    "filePath": "/Users/aidancunniffe/Developer/openapi-optic/poc-governance-tools/projects/snyk-rules/end-end-tests/api-standards/components/parameters/version.yaml",
    "preview": "",
    "startLine": 1,
    "startPosition": 0,
  },
  undefined,
  undefined,
  undefined,
  undefined,
  Object {
    "endLine": 12,
    "endPosition": 313,
    "filePath": "/Users/aidancunniffe/Developer/openapi-optic/poc-governance-tools/projects/snyk-rules/end-end-tests/api-standards/components/responses/403.yaml",
    "preview": "",
    "startLine": 5,
    "startPosition": 228,
  },
  undefined,
  undefined,
  Object {
    "endLine": 12,
    "endPosition": 374,
    "filePath": "/Users/aidancunniffe/Developer/openapi-optic/poc-governance-tools/projects/snyk-rules/end-end-tests/api-standards/components/responses/400.yaml",
    "preview": "",
    "startLine": 6,
    "startPosition": 297,
  },
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  Object {
    "endLine": 12,
    "endPosition": 561,
    "filePath": "/Users/aidancunniffe/Developer/openapi-optic/poc-governance-tools/projects/snyk-rules/end-end-tests/api-standards/components/responses/500.yaml",
    "preview": "",
    "startLine": 8,
    "startPosition": 494,
  },
  undefined,
  Object {
    "endLine": 12,
    "endPosition": 313,
    "filePath": "/Users/aidancunniffe/Developer/openapi-optic/poc-governance-tools/projects/snyk-rules/end-end-tests/api-standards/components/responses/403.yaml",
    "preview": "",
    "startLine": 5,
    "startPosition": 228,
  },
  undefined,
  undefined,
  undefined,
  undefined,
  Object {
    "endLine": 12,
    "endPosition": 534,
    "filePath": "/Users/aidancunniffe/Developer/openapi-optic/poc-governance-tools/projects/snyk-rules/end-end-tests/api-standards/components/responses/404.yaml",
    "preview": "",
    "startLine": 8,
    "startPosition": 467,
  },
  undefined,
  Object {
    "endLine": 12,
    "endPosition": 472,
    "filePath": "/Users/aidancunniffe/Developer/openapi-optic/poc-governance-tools/projects/snyk-rules/end-end-tests/api-standards/components/responses/400.yaml",
    "preview": "",
    "startLine": 7,
    "startPosition": 379,
  },
  Object {
    "endLine": 12,
    "endPosition": 364,
    "filePath": "/Users/aidancunniffe/Developer/openapi-optic/poc-governance-tools/projects/snyk-rules/end-end-tests/api-standards/components/responses/404.yaml",
    "preview": "",
    "startLine": 6,
    "startPosition": 287,
  },
  undefined,
  undefined,
  Object {
    "endLine": 31,
    "endPosition": 172,
    "filePath": "/Users/aidancunniffe/Developer/openapi-optic/poc-governance-tools/projects/snyk-rules/end-end-tests/api-standards/components/errors.yaml",
    "preview": "",
    "startLine": 5,
    "startPosition": 96,
  },
  undefined,
]
`;
